packages/flutter/lib/src/cupertino/sheet.dart

```diff
@@ -0,0 +1,349 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/services.dart';
+import 'package:flutter/widgets.dart';
+
+import 'colors.dart';
+import 'route.dart';
+
+// Offset from offscreen below to stopping below the top of the screen.
+final Animatable<Offset> _kBottomUpTween = Tween<Offset>(
+  begin: const Offset(0.0, 1.0),
+  end: const Offset(0.0, 0.08),
+);
+
+final Animatable<Offset> _kFullBottomUpTween = Tween<Offset>(
+  begin: const Offset(0.0, 1.0),
+  end: const Offset(0.0, -0.02),
+);
+
+// Offset slightly up when a sheet gets covered by another sheet.
+final Animatable<Offset> _kMidUpTween = Tween<Offset>(
+  begin: Offset.zero,
+  end: const Offset(0.0, -0.005),
+);
+
+// Offset from top of screen to slightly down when covered by a sheet.
+final Animatable<Offset> _kTopDownTween = Tween<Offset>(
+  begin: Offset.zero,
+  end: const Offset(0.0, 0.07),
+);
+
+// Amount the sheet in the background scales down. Found by measuring the width
+// of the sheet in the background and comparing against the screen width. The
+// scale transition will go from a default of 1.0 to 1.0 - _kSheetScaleFactor.
+const double _kSheetScaleFactor = 0.0835;
+
+final Animatable<double> _kScaleTween = Tween<double>(begin: 1.0, end: 1.0 - _kSheetScaleFactor);
+
+/// Docs placeholder
+Future<T?> showCupertinoSheet<T>({
+  required BuildContext context,
+  required WidgetBuilder pageBuilder,
+  bool useRootNavigator = true,
+  bool useNestedNavigation = false,
```

Comment:
Can you try to get state restoration working here? Whether in this code or in the example code. I really wonder what this would look like in a fully-fledged production navigation scenario, in order to decide what to do here.

Reply:
Sorry, I talked about how iOS does it [outside this comment](https://github.com/flutter/flutter/pull/157568#issuecomment-2480036684). TLDR is on iOS when you add a sheet it automatically adds something like a Navigator to the content so nested navigation happens by default. If you want to do unnested navigation you have to pass a callback to the sheet content that will trigger navigation outside of the scope of the sheet.

Reply:
ah I see, I think the native ios behavior is the same as if we add navigator by default. If user want to add a unnest page they need to do Navigator .push with the context above the cupertino sheet.

I don't have strong opinion on whether we need to follow them. your call.